internal/rbd/replicationcontrollerserver.go

```diff
@@ -0,0 +1,357 @@
+// If the image is already secondary it will return success.
+func (cs *ControllerServer) DemoteVolume(ctx context.Context,
+	req *replication.DemoteVolumeRequest,
+) (*replication.DemoteVolumeResponse, error) {
+	rbdVol, cr, err := cs.getVolumeFromID(ctx, req.GetVolumeId(), req.GetSecrets())
+	defer cs.cleanup(rbdVol, cr)
+	if err != nil {
+		return nil, err
+	}
+
+	mirroringInfo, err := rbdVol.getImageMirroringInfo()
+	if err != nil {
+		util.ErrorLog(ctx, err.Error())
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	if mirroringInfo.State != librbd.MirrorImageEnabled {
+		return nil, status.Errorf(codes.InvalidArgument, "mirroring is not enabled on %s, image is in %d Mode", rbdVol.VolID, mirroringInfo.State)
+	}
+
+	// demote image to secondary
+	if mirroringInfo.Primary {
+		err = rbdVol.demoteImage()
+		if err != nil {
+			util.ErrorLog(ctx, err.Error())
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+	}
+	return &replication.DemoteVolumeResponse{}, nil
+}
+
+// ResyncVolume extracts the RBD volume information from the volumeID, If the
+// image is present, mirroring is enabled and the image is in demoted state.
+// If yes it will resync the image to correct the split-brain.
+func (cs *ControllerServer) ResyncVolume(ctx context.Context,
+	req *replication.ResyncVolumeRequest,
+) (*replication.ResyncVolumeResponse, error) {
+	rbdVol, cr, err := cs.getVolumeFromID(ctx, req.GetVolumeId(), req.GetSecrets())
+	defer cs.cleanup(rbdVol, cr)
+	if err != nil {
+		return nil, err
+	}
+
+	mirroringInfo, err := rbdVol.getImageMirroringInfo()
+	if err != nil {
+		// TODO: check if the image is not found return valid error message
+		// in case of Resync the image will get deleted and gets recreated and
+		// it takes time for this operation
+		util.ErrorLog(ctx, err.Error())
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	if mirroringInfo.State != librbd.MirrorImageEnabled {
+		return nil, status.Error(codes.InvalidArgument, "image mirroring is not enabled")
+	}
+
+	// return error if the image is still primary
+	if mirroringInfo.Primary {
+		return nil, status.Error(codes.InvalidArgument, "image is in primary state")
+	}
+
+	err = rbdVol.resyncImage()
+	if err != nil {
+		util.ErrorLog(ctx, err.Error())
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	// TODO: check the image state and return its ready to use or not
+	mirrorStatus, err := rbdVol.getImageMirroingStatus()
+	if err != nil {
+		// the image gets recreated after issusing resync in that case return
+		// volume as not ready.
+		if errors.Is(err, ErrImageNotFound) {
+			resp := &replication.ResyncVolumeResponse{
+				Ready: false,
+			}
+			return resp, nil
+		}
+		util.ErrorLog(ctx, err.Error())
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	ready := false
+	state := imageMirroringState(mirrorStatus.State)
+	if state == upAndStopped || state == upAndReplaying {
```

Comment:
These are the only two states that are defined in the top of the file, can there be others?

Probably the `ResyncVolumeResponse` should contain a `Status` free-form field so that users can see more details about the status? A single boolean value (not)ready is useful for automation, but we also need to assist users that have to debug (configuration) issues. Or include some events, somehow, so that there is a log tied with the CR with changes in the mirroring state and timestamps (like events of a PVC or Pod).

Reply:
Yes good to add it to spec. AFAIK only these two are the state which we need to return success the above state ensures that the image is in good condition to get used.